### PR TITLE
Add `maximumDistance` option

### LINKED ISSUE #14: parameter for maximum distance
When testing many for or sorting by similarity, often we are happy to short-circuit on any pair with a distance greater than `max`.

`leven(a, b, 3)`

At scale, given that the least similar strings are also very expensive to compute, there is a big potential savings.  (For one, just comparing length is enough to discard many candidates.)

(For such pairs, the distance can be `null`, `max` or `max + 1` - I don't have a strong opinion on which is best.)

### PR DESCRIPTION
**Pull request by reganmeloche**
Adds an optional parameter for maximum distance and short circuits if the difference in length is greater than or equal to that value. Includes tests

Fixes #14

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ declare const leven: {
 	//=> 2
 	```
 	*/
-	(left: string, right: string): number;
+	(left: string, right: string, options?: any): number;
```
Comment (sindresorhus) [inline index.d.ts]: The options should be properly typed and documented here. See: https://github.com/sindresorhus/typescript-definition-style-guide#documentation

File: test.js [hunk 2]
```diff
@@ -16,4 +16,7 @@ test('main', t => {
 	t.is(leven('levenshtein', 'frankenstein'), 6);
 	t.is(leven('distance', 'difference'), 5);
 	t.is(leven('因為我是中國人所以我會說中文', '因為我是英國人所以我會說英文'), 2);
+	t.is(leven('123', '12345678', {fakeValue: 1}), 5);
+	t.is(leven('123', '12345678', {maximumDistance: 3}), 3);
+	t.is(leven('123', '12345678', {maximumDistance: 6}), 5);
```
Comment (sindresorhus) [inline test.js]: These should be in a separate test function.

**Comment by sindresorhus** [comment]
The option needs to be documented in the readme. (Example: https://github.com/sindresorhus/boxen#api)

**Comment by Yomguithereal** [comment]
It is a bit weird to only use the length comparison if you want to implement a Levenshtein with max distance. Indeed, usually people tend to apply Ukkonen's optimization for this. You can find such implementation [here](https://github.com/Yomguithereal/talisman/blob/master/src/metrics/distance/levenshtein.js#L230) which is mostly similar to `leven`'s one.

**Comment by MaximDevoir** [comment]
@Yomguithereal I benchmarked your implementation and it is very fast.

```
    89,217 op/s » leven
   225,085 op/s » leven maxDistance:5
   120,088 op/s » leven maxDistance:10
    96,507 op/s » talisman
  7,187,299 op/s » talisman maxDistance:5
  7,288,593 op/s » talisman maxDistance:10
```

Would you be able to build on top of #15 or #16?

**Comment by Yomguithereal** [comment]
@MaximDevoir basically all the heuristics existing in Talisman about the non-max Levenshtein were included into `leven`. Both implementation only differ in that Talisman's one is able to process arbitrary sequences as in arrays which accounts for the slight performance issue since Talisman's one is polymorphic.

But there is a faster implementation using loop unrolling [here](https://github.com/gustf/js-levenshtein) (there is also a lib using native bindings and SIMD which is even faster but won't work in browsers, obviously).

Now, concerning the max distance implementation: it should really be another function and not an option of the main one because there is a performance cost that will be paid for cases when you don't want to impose a max distance. What's more, [gustf](https://github.com/gustf) loop unrolling does not yield perf improvements on the max distance versions.

This said, I have no objection if you want to copy code from Talisman into a max distance function (separated from the main one) so it can be exported by `leven`. But the decision really should be @sindresorhus's

**Comment by MaximDevoir** [comment]
Whoops, I made a mistake big mistake in the my benchmark test. I've since fixed the benchmark test and have updated [this comment](https://github.com/sindresorhus/leven/pull/15#issuecomment-562769225) to include the accurate benchmark results. I also uploaded my source code for the benchmark [here](https://github.com/sindresorhus/leven/pull/16/commits/5ac4781eee044b36effdefea957cfacca4a154a6).

**Comment by sindresorhus** [comment]
@reganmeloche Can you address my feedback?

**Comment by sindresorhus** [comment]
I think a simple implementation of this is fine. If you want super optimized performance, JavaScript is not the right choice anyway.

**Comment by sindresorhus** [comment]
Bump

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/leven/issues/14#issuecomment-654360372

**Pull request closed without merging**